File: src_py/formulations/intermediate.py
import math

import igraph as ig

from constants import eps
from formulations.formulation_abstract import Formulation
from utils.graph_utils import simple_cycles_limited_length, simple_cycles_parallel
from utils.transport import Input, Output
from utils.utils import get_remaining_time, setdiff


class Intermediate(Formulation):
    def __init__(self, input_data: Input):
        self.start_time = input_data.start_time
        self.timed_out = False

        self.m = input_data.solver_instance("basic")
        self.max_cycle_length = input_data.cycle_length
        self.max_chain_length = input_data.chain_length
        self.graph = graph = input_data.graph

        self.ndds = input_data.ndds
        self.weights = input_data.weights

        self.forbidden_nodes = input_data.forbidden_nodes

        n = graph.vcount()
        self.n = n
        vertices = graph.vs
        edges = graph.es
        self.edge_tuples = [e.tuple for e in edges]
        self.vertex_indices = [v.index for v in vertices]
        self.set_neighbors()

    def solve(self):
        self.prepare_cycles()
        if self.timed_out:
            return

        self.create_aux_graph()
        self.set_lp()
        self.m.set_callbacks(self.callback, lazy=True, cut=True)

        obj_val = -1
        optimal = False
        cuts_correct = False
        while not cuts_correct:
            remaining_time = get_remaining_time(self.start_time)
            if remaining_time > 0.1:
                obj_val, optimal = self.m.solve(remaining_time)
            if not optimal:
                return None
            # cuts_correct = True
            cuts_correct = self.check_cuts()
        return self.get_output(obj_val, optimal), self.__class__

    def set_lp(self):
        n_cycles = len(self.cycles)
        ndds = self.ndds
        self.patients = setdiff(self.vertex_indices, ndds)

        # print("Adding variables")
        self.x = self.m.add_vars(self.edge_tuples)
        self.in_flow = self.m.add_vars(self.vertex_indices)
        self.out_flow = self.m.add_vars(self.vertex_indices)
        self.z = self.m.add_vars(list(range(n_cycles)))

        # print("Adding patient constraints")
        for k in self.patients:
            expr = self.in_flow[k] + -1 * self.out_flow[k]
            if k in self.forbidden_nodes:
                self.m.add_constr_eq(expr, 0)
            else:
                self.m.add_constr_ge(expr, 0)
            cycles_k = self.cycles_part_of[k]
            var_list = [self.z[c] for c in cycles_k]
            var_list.append(self.in_flow[k])
            self.m.add_constr_le(self.m.quick_sum(var_list), 1)

        # print("Adding NDD variables and constraints")
        for n in ndds:
            self.m.add_constr_eq(self.in_flow[n], 0)

        if self.max_chain_length < self.n-1:  # bounded chains
            x_ndds = [None]*len(ndds)
            in_ndds = [None]*len(ndds)
            out_ndds = [None]*len(ndds)
            for n in ndds:
                x_ndds[n] = self.m.add_vars(self.edge_tuples)
                in_ndds[n] = self.m.add_vars(self.vertex_indices)
                out_ndds[n] = self.m.add_vars(self.vertex_indices)

                self.m.add_constr_le(self.m.quick_sum(x_ndds[n].values()), self.max_chain_length)
                for k in self.vertex_indices:
                    self.m.add_constr_eq(self.m.quick_sum([x_ndds[n][(i, k)] for i in self.in_neighbors[k]]) + -1*in_ndds[n][k], 0)
                    self.m.add_constr_eq(self.m.quick_sum([x_ndds[n][(k, j)] for j in self.out_neighbors[k]]) + -1*out_ndds[n][k], 0)
                for p in self.patients:
                    self.m.add_constr_le(out_ndds[n][p] + -1*in_ndds[n][p], 0)
            for e in self.edge_tuples:
                self.m.add_constr_eq(self.m.quick_sum([x_ndds[n][e] for n in ndds]) + -1*self.x[e], 0)
        elif len(ndds) == 0:
            for e in self.edge_tuples:
                self.m.add_constr_eq(self.x[e], 0)

        # print("Setting objective and flow constraints")
        obj_list = [None]*n_cycles
        for c in range(n_cycles):
            obj_list[c] = self.z[c] * self.cycle_weights[c]

        for k in self.vertex_indices:
            expr_in = self.m.quick_sum([self.x[(i, k)] for i in self.in_neighbors[k]]) + -1*self.in_flow[k]
            expr_out = self.m.quick_sum([self.x[(k, j)] for j in self.out_neighbors[k]]) + -1*self.out_flow[k]
            self.m.add_constr_eq(expr_in, 0)
            self.m.add_constr_eq(expr_out, 0)
            for j in self.out_neighbors[k]:
                obj_list.append(self.x[(k, j)] * self.weights[k, j])

        self.m.set_objective_list(obj_list)
        return

    def set_neighbors(self):
        n = self.n

        self.in_neighbors = [None]*n
        self.out_neighbors = [None]*n
        for k in self.vertex_indices:
            self.in_neighbors[k] = self.graph.neighbors(k, mode=ig.IN)
            self.out_neighbors[k] = self.graph.neighbors(k, mode=ig.OUT)

        self.in_neighbors_tuples = [None] * n
        self.out_neighbors_tuples = [None] * n
        for v in range(n):
            self.in_neighbors_tuples[v] = []
            self.out_neighbors_tuples[v] = []
        for e in self.edge_tuples:
            orig = e[0]
            dest = e[1]
            self.in_neighbors_tuples[dest] += [e]
            self.out_neighbors_tuples[orig] += [e]

    def prepare_cycles(self):
        print("Preparing cycles")
        remaining_time = get_remaining_time(self.start_time)
        max_cycle_length = min(self.max_cycle_length, self.n - len(self.ndds))
        cycles = simple_cycles_limited_length(self.graph, max_cycle_length, remaining_time)
        # cycles = simple_cycles_parallel(self.graph, max_cycle_length)
        if cycles is None:
            self.timed_out = True
            return
        num_cycles = len(cycles)
        print("Found %s cycles of length at most %s" % (num_cycles, max_cycle_length))
        w = [0]*num_cycles
        weights = self.weights
        cycles_part_of = [[] for _ in range(self.n)]
        for i, cycle in enumerate(cycles):
            n = len(cycle)
            w[i] = sum(weights[cycle[j], cycle[j+1]] for j in range(n-1))
            w[i] += weights[cycle[n-1], cycle[0]]

            for j in cycle:
                cycles_part_of[j].append(i)
        self.cycles = cycles
        self.cycle_weights = w
        self.cycles_part_of = cycles_part_of

    def create_aux_graph(self):
        cb_graph = self.graph.copy()
        caps = dict()
        n = self.n
        cb_graph.add_vertex()
        new_edges = [None]*len(self.ndds)
        for j, i in enumerate(self.ndds):
            new_edges[j] = (n, i)
            caps[n, i] = 1
            caps[i, n] = math.inf
        cb_graph.add_edges(new_edges)
        self.cb_graph = cb_graph
        self.cb_tuples = [e.tuple for e in cb_graph.es]
        self.added_sets = dict()
        self.caps = caps

    def callback(self, data):
        if len(self.ndds) == 0:
            return
        x_vals = self.m.callback.get_val(self.x, data)
        in_vals = self.m.callback.get_val(self.in_flow, data)
        pos_index = []
        pos_values = []
        for i in self.patients:
            if in_vals[i] > eps:
                pos_index.append(i)
                pos_values.append(in_vals[i])
        edges_sets = self.find_sets(x_vals, pos_index, pos_values)
        for i, edges_set in edges_sets:
            expr = self.m.quick_sum([self.x[e] for e in edges_set]) + -1*self.in_flow[i]
            self.m.callback.add_constr_ge(expr, 0, data)

    def check_cuts(self):
        if len(self.ndds) == 0:
            return True
        x_vals = self.m.get_val(self.x)
        in_vals = self.m.get_val(self.in_flow)
        pos_index = []
        pos_values = []
        for i in self.patients:
            if in_vals[i] > eps:
                pos_index.append(i)
                pos_values.append(in_vals[i])
        edges_sets = self.find_sets(x_vals, pos_index, pos_values)
        for i, edges_set in edges_sets:
            expr = self.m.quick_sum([self.x[e] for e in edges_set]) + -1*self.in_flow[i]
            self.m.add_constr_ge(expr, 0)
        return len(edges_sets) == 0

    def find_sets(self, x_vals, nodes, in_vals):
        n, ndds, graph, cb_graph, caps = self.n, self.ndds, self.graph, self.cb_graph, self.caps
        edges_sets = []
        for i in ndds:
            x_vals[(n, i)] = 1
        edge_tuples = self.cb_tuples
        values = [x_vals[e] for e in edge_tuples]
        cb_graph.es["capacity"] = values
        for v, in_value in zip(nodes, in_vals):
            mincut = cb_graph.mincut(n, v, "capacity")
            flow = mincut.value
            part2 = mincut[1]
            if flow >= in_value-eps:
                continue
            edges_set = []
            for j in setdiff(part2, ndds):
                in_neighbors = self.in_neighbors[j]
                vec = setdiff(in_neighbors, part2)
                for k in vec:
                    edges_set.append((k, j))
            edges_set.sort()  # to avoid repeating constraints
            edges_sets.append((v, edges_set))
        return edges_sets

    def get_output(self, value, optimal):
        match_edges = self.m.get_val(self.x)
        match_cycles = self.m.get_val(self.z)
        graph_cycles = self.cycles
        output = Output(match_edges, match_cycles, graph_cycles, value, optimal)
        return output

File: src_jl/intermediate.jl
using LightGraphsFlows
using SharedArrays
using JuMP
include("transport.jl")
const eps=1e-8

struct Intermediate
    m::Model
    x::JuMP.Containers.DenseAxisArray
    z::Array{VariableRef,1}
    in_flow::Array{VariableRef,1}
    out::Array{VariableRef,1}
    n::Int64
    max_cycle_length::Int64
    max_chain_length::Any
    graph::DiGraph
    graph_edges::Any
    index_to_edge::Dict{Any,Any}
    ndds::Array{Int64, 1}
    cb_graph::DiGraph
    caps::Array{Float64,2}
    graph_cycles::Any
    time_param::String
    time_factor::Float64
    start_time::Float64

    function Intermediate(input_data::Input)
        start_time = input_data.start_time
        graph, weights, ndds, max_cycle_length, max_chain_length, solver_instance = input_data.graph, input_data.weights, input_data.ndds, input_data.max_cycle_length, input_data.max_chain_length, input_data.solver_instance
        m = Model(solver_instance)
        n = nv(graph)
        index_to_edge = Dict()
        graph_edges = collect(edges(graph))
        for e in graph_edges
            index_to_edge[(e.src, e.dst)] = e
        end
        cycle_weights, c = get_cycles(graph, max_cycle_length, weights)
        num_cycles = size(c)[1]
        @variable(m, 1 >= x[graph_edges] >= 0, Int)
        @variable(m, 1 >= in_flow[1:n] >= 0, Int)
        @variable(m, 1 >= out[1:n] >= 0, Int)
        @variable(m, 1 >= z[1:num_cycles] >= 0, Int)

        @objective(m, Max, sum(x[Edge((i, j))] * weights[i, j] for i in 1:n, j in outneighbors(graph, i)) + sum(z[i] * cycle_weights[i] for i in 1:num_cycles))
        @constraint(m, c1[k=1:n], sum(x[Edge((i,k))] for i=inneighbors(graph, k)) == in_flow[k])
        @constraint(m, c2[k=1:n], sum(x[Edge((k,j))] for j=outneighbors(graph, k)) == out[k])
        @constraint(m, c3[k=setdiff(1:n, ndds)], out[k] <= in_flow[k])
        @constraint(m, c4[k=ndds], in_flow[k] == 0)
        @constraint(m, c5[k=setdiff(1:n, ndds)], in_flow[k] <= 1-sum(z[j] for j in cycles_part_of(c, k)))
        if length(ndds) == 0
            @constraint(m, d[e=graph_edges], x[e] == 0)
        end
        
        if max_chain_length < n-1
            @variable(m, 1 >= x_ndds[graph_edges, ndds] >= 0, Int)
            @variable(m, 1 >= in_ndds[1:n, ndds], Int)
            @variable(m, out_ndds[1:n, ndds], Int)

            @constraint(m, c6[e=graph_edges], sum(x_ndds[e, k] for k in ndds) == x[e])
            @constraint(m, c7[k=ndds], sum(x_ndds[e, k] for e in graph_edges) <= max_chain_length)
            @constraint(m, c8[k=1:n, ndd=ndds], sum(x_ndds[Edge((i,k)), ndd] for i=inneighbors(graph, k)) == in_ndds[k, ndd])
            @constraint(m, c9[k=1:n, ndd=ndds], sum(x_ndds[Edge((k,j)), ndd] for j=outneighbors(graph, k)) == out_ndds[k, ndd])
            @constraint(m, c10[k=setdiff(1:n, ndds), ndd=ndds], out_ndds[k, ndd] <= in_ndds[k, ndd])
        end
        cb_graph = deepcopy(graph)
        caps = zeros(Float64, n+1, n+1)
        add_vertices!(cb_graph, 1)
        for i in ndds
            add_edge!(cb_graph, n+1, i)
            caps[n+1, i] = 1
            caps[i, n+1] = Inf
        end
        println("Number of variables: ", length(all_variables(m)))
        # println("Number of constraints: ", num_constraints(m, Any, Any))
        time_param = input_data.time_param
        time_factor = input_data.time_factor
        new(m, x, z, in_flow, out, n, max_cycle_length, max_chain_length, graph, graph_edges, index_to_edge, ndds, cb_graph, caps, c, time_param, time_factor, start_time)
    end
end

function solve(problem::Intermediate, signal::SharedArray{Bool,1}=SharedArray{Bool}(1))
    output = Output()

    function general_callback(cb_data, MOI_instance)
        if length(problem.ndds) == 0
            return
        end
        if signal[1]
            println("Stopping solve")
            get_best_solution(problem, output, cb_data)
            con = @build_constraint(0 >= 1)
            MOI.submit(problem.m, MOI_instance(cb_data), con)
        end

        x_vals = callback_value.(Ref(cb_data), problem.x)
        in_vals = callback_value.(Ref(cb_data), problem.in_flow)
        pos_index = []
        for i in 1:problem.n
            if in_vals[i] > 0
                push!(pos_index, i)
            end
        end
        edges_sets = find_sets(x_vals, pos_index, problem)
        for pair in edges_sets
            i, edges_set = pair
            con = @build_constraint(sum(problem.x[e] for e in edges_set) >= problem.in_flow[i])
            MOI.submit(problem.m, MOI_instance(cb_data), con)
        end
    end

    function lazy_callback(cb_data)
        general_callback(cb_data, MOI.LazyConstraint)
    end
    MOI.set(problem.m, MOI.LazyConstraintCallback(), lazy_callback)

    function user_cut_callback(cb_data)
        general_callback(cb_data, MOI.UserCut)
    end
    MOI.set(problem.m, MOI.UserCutCallback(), user_cut_callback)

    tf = time()
    remaining_time = max(timeout-(tf-problem.start_time), 1)*problem.time_factor
    remaining_time = convert(Int32, round(remaining_time, digits=0))
    set_parameter(problem.m, problem.time_param, remaining_time)

    optimize!(problem.m)
    if termination_status(problem.m) == MOI.TIME_LIMIT
        println("Intermediate timed out")
        return nothing
    end
    return get_output(problem)
end

function get_output(problem::Intermediate)
    match_vertices = JuMP.value.(problem.in_flow)
    match_edges = JuMP.value.(problem.x)
    match_cycles = JuMP.value.(problem.z)
    graph_cycles = problem.graph_cycles
    value = JuMP.objective_value(problem.m)
    optimal = true
    output = Output(match_vertices, match_edges, match_cycles, graph_cycles, value, optimal)
    return output
end

function get_best_solution(problem::Intermediate, output::Output, cb_data::Any)
    return
end

"""
get_cycles
Returns list of weights for each cycle of length max k and list of said cycles.
"""
function get_cycles(graph::DiGraph, max_cycle_length::Int64, weights::Array{Float64,2})
    max_cycles = 1e8
    Ck = simplecycles_limited_length(graph, max_cycle_length, max_cycles)
    num_cycles = size(Ck)[1]
    println("Found $num_cycles cycles of length at most $max_cycle_length")
    w = zeros(Float64, num_cycles)
    for (i, cycle) in enumerate(Ck)
        n = size(cycle)[1]
        if n == 1
            w[i] = weights[cycle[1], cycle[1]]
        else
            w[i] = sum(weights[cycle[j], cycle[j+1]] for j in 1:n-1)
            w[i] += weights[cycle[n], cycle[1]]
        end
    end
    return (w, Ck)
end

"""
cycles_part_of
Returns list of indices corresponding to cycles that node k is a part of given
    the list C of cycles
"""
function cycles_part_of(C::Array{Array{Int64,1},1}, k::Int64)
    indices = zeros(Int64, 0)
    for (i,c) in enumerate(C)
        if k in c
            append!(indices, i)
        end
    end
    return indices
end

function find_sets(x_vals::JuMP.Containers.DenseAxisArray, pos_index::Array{Any, 1}, problem::Intermediate)
    n, ndds, graph, cb_graph, caps, index_to_edge = problem.n, problem.ndds, problem.graph, problem.cb_graph, problem.caps, problem.index_to_edge
    cb_caps = deepcopy(caps)
    for e in problem.graph_edges
        cb_caps[e.src,e.dst] = x_vals[e]
    end
    edges_sets = []
    for i in setdiff(pos_index, ndds)
        part1, part2, flow = minimum_cut(cb_graph, n+1, i, cb_caps)
        if flow <= 1-eps
            edges_set = []
            for j in setdiff(part2, ndds)
                vec = setdiff(inneighbors(graph, j),part2)
                for k in vec
                    push!(edges_set, index_to_edge[(k,j)])
                end
            end
            push!(edges_sets, (i, edges_set))
        end
    end
    return edges_sets
end

function minimum_cut(
        flow_graph::DiGraph,                 # the input graph
        source::Int64,                       # the source vertex
        target::Int64,                       # the target vertex
        capacity_matrix::Array{Float64,2},   # edge flow capacities
        # algorithm::AbstractFlowAlgorithm   # keyword argument for algorithm
    )
    flow, flow_matrix = maximum_flow(flow_graph, source, target, capacity_matrix, EdmondsKarpAlgorithm())
    # flow, flow_matrix = maximum_flow(flow_graph, source, target, capacity_matrix, PushRelabelAlgorithm())
    residual_matrix = zeros(nv(flow_graph),nv(flow_graph))
    for edge in edges(flow_graph)
        residual_matrix[edge.src,edge.dst] = max(0.0, capacity_matrix[edge.src,edge.dst] - flow_matrix[edge.src,edge.dst])
    end
    part1 = typeof(source)[]
    queue = [source]
    while !isempty(queue)
        node = pop!(queue)
        push!(part1, node)
        dests = [dst for dst in 1:nv(flow_graph) if residual_matrix[node,dst] > 0.0 && dst ∉ part1]
        append!(queue, dests)
    end
    part2 = [node for node in 1:nv(flow_graph) if node ∉ part1]
    return (part1, part2, flow)
end
